src/app/services/JoinerService.ts

```diff
@@ -51,68 +90,63 @@ export class JoinerService {
-        } else if (user.id === joiner.get().creator.id) {
-            return MGPValidation.SUCCESS;
         } else {
-            joinerList[joinerList.length] = user;
-            await this.joinerDAO.update(partId, { candidates: joinerList });
+            if (joiner.get().creator.id !== user.id) {
+                // Only add actual candidates to the game, not the creator
+                await this.joinerDAO.addCandidate(partId, user);
+            }
             return MGPValidation.SUCCESS;
         }
     }
-    public async cancelJoining(user: MinimalUser): Promise<void> {
+    public async removeCandidate(candidate: MinimalUser): Promise<void> {
+        assert(this.observedJoinerId != null, 'cannot remove candidate if not observing a joiner');
+        return this.joinerDAO.removeCandidate(Utils.getNonNullable(this.observedJoinerId), candidate);
+    }
+    public async cancelJoining(): Promise<void> {
         display(JoinerService.VERBOSE,
-                'JoinerService.cancelJoining(' + user.name + '); this.observedJoinerId = ' + this.observedJoinerId);
+                'JoinerService.cancelJoining(); this.observedJoinerId = ' + this.observedJoinerId);
 
         if (this.observedJoinerId == null) {
             throw new Error('cannot cancel joining when not observing a joiner');
         }
+        const user: MinimalUser = this.connectedUserService.user.get().toMinimalUser();
         const joinerOpt: MGPOptional<Joiner> = await this.joinerDAO.read(this.observedJoinerId);
         if (joinerOpt.isAbsent()) {
             // The part does not exist, so we can consider that we succesfully cancelled joining
             return;
         } else {
             const joiner: Joiner = joinerOpt.get();
-            const candidates: MinimalUser[] = ArrayUtils.copyImmutableArray(joiner.candidates);
-            const indexLeaver: number = candidates.findIndex((u: MinimalUser) => u.id === user.id);
             let chosenOpponent: MinimalUser | null = joiner.chosenOpponent;
             let partStatus: number = joiner.partStatus;
-            candidates.splice(indexLeaver, 1); // remove player from candidates list
             if (chosenOpponent?.id === user.id) {
-                // if the ChosenOpponent left, we're back to initial part creation
+                // if the chosenOpponent leave, we're back to initial part creation
                 chosenOpponent = null;
                 partStatus = PartStatus.PART_CREATED.value;
-            } else if (indexLeaver === -1) {
-                throw new Error('someone that was not candidate nor ChosenOpponent just left the chat: ' + user.name);
             }
             const update: Partial<Joiner> = {
                 chosenOpponent: chosenOpponent,
                 partStatus,
-                candidates,
             };
-            return this.joinerDAO.update(this.observedJoinerId, update);
+            await this.joinerDAO.update(this.observedJoinerId, update);
```

Comment:
if the user who leave is not chosenOpponent, one of this modifications will be useless and should be skipped !

Reply:
vrai, adapté